build_tools/circle/build_doc.sh

```diff
@@ -2,18 +2,82 @@
 set -x
 set -e
 
-# Introspect the commit to know whether or not we should skip building the
-# documentation: a pull request that does not change any file in doc/ or
-# examples/ folder should be skipped unless the "[doc: build]" is found the
-# commit message.
-BUILD_DOC=`python build_tools/circle/check_build_doc.py`
-echo -e $BUILD_DOC
-if [[ $BUILD_DOC == "SKIP:"* ]]; then
-    touch ~/log.txt  # the "test" segment needs that file
+# Decide what kind of documentation build to run, and run it.
+#
+# If the last commit message has a "[doc skip]" marker, do not build
+# the doc. On the contrary if a "[doc build]" marker is found, build the doc
+# instead of relying on the subsequent rules.
+#
+# We always build the documentation for jobs that are not related to a specific
+# PR (e.g. a merge to master or a maintenance branch).
+#
+# If this is a PR, do a full build if there are some files in this PR that are
+# under the "doc/" or "examples/" folders, otherwise perform a quick build.
+#
+# If the inspection of the current commit fails for any reason, the default
+# behavior is to quick build the documentation.
+
+get_build_type() {
+	if [ -z "$CIRCLE_SHA1" ]
+	then
+		echo SKIP: undefined CIRCLE_SHA1
+		return
+	fi
+	commit_msg=$(git log --format=%B -n 1 $CIRCLE_SHA1)
+	if [ -z "$commit_msg" ]
+	then
+		echo QUICK BUILD: failed to inspect commit $CIRCLE_SHA1
+		return
+	fi
+	if [[ "$commit_msg" =~ '.*[doc skip].*' ]]
+	then
+		echo SKIP: [doc skip] marker found
+		return
+	fi
+	if [[ "$commit_msg" =~ '.*[doc build].*' ]]
+	then
+		echo BUILD: [doc build] marker found
+		return
+	fi
+	if [ -z "$CI_PULL_REQUEST" ]
+	then
+		echo BUILD: not a pull request
+		return
+	fi
+	git_range="origin/master...$CIRCLE_SHA1"
+	git fetch origin master >&2
+	filenames=$(git diff --name-only $git_range)
+	if [ -z "$filenames" ]
+	then
+		echo QUICK BUILD: failed to get changed filenames for $git_range
+		return
+	fi
+	if echo "$filenames" | grep -q -e ^examples/ -e ^doc/
```

Comment:
I got tripped up by this one a few times: since you have `set -e` a grep without a match will have an exit code 1 and stop the script.

I have to say I don't know what is the best way to handle this so I used things like this in flake8_diff.sh:
```
echo $filenames | grep bla || echo no_match
```